HARK/ConsumptionSaving/ConsLabeledModel.py

```diff
@@ -0,0 +1,663 @@
+        # evaluate the value function at that point, so we must fill out the data
+        if self.BoroCnstArt is None or self.BoroCnstArt <= self.BoroCnstNat:
+            self.m_nrm_min = self.BoroCnstNat
+            self.nat_boro_cnst = True
+
+            self.borocnst = xr.Dataset(
+                coords={"mNrm": self.m_nrm_min, "aNrm": self.m_nrm_min},
+                data_vars={
+                    "cNrm": 0.0,
+                    "v": -np.inf,
+                    "v_inv": 0.0,
+                    "reward": -np.inf,
+                    "marginal_reward": np.inf,
+                    "v_der": np.inf,
+                    "v_der_inv": 0.0,
+                },
+            )
+
+        elif self.BoroCnstArt > self.BoroCnstNat:
+            self.m_nrm_min = self.BoroCnstArt
+            self.nat_boro_cnst = False
+
+            self.borocnst = xr.Dataset(
+                coords={"mNrm": self.m_nrm_min, "aNrm": self.m_nrm_min},
+                data_vars={"cNrm": 0.0},
+            )
+
+    def create_post_state(self):
+
+        if self.nat_boro_cnst:
+            # don't include natural borrowing constraint
+            a_grid = self.aXtraGrid + self.m_nrm_min
+        else:
+            # include artificial borrowing constraint
+            a_grid = np.append(0.0, self.aXtraGrid) + self.m_nrm_min
+
+        aVec = xr.DataArray(
+            a_grid,
+            name="aNrm",
+            dims=("aNrm"),
+            attrs={"long_name": "savings", "state": True},
+        )
+        post_state = xr.Dataset({"aNrm": aVec})
+
+        self.post_state = post_state
+
+    def state_transition(self, s=None, a=None, params=None):
+        """
+        state to post_state transition
+        """
+        ps = {}  # pytree
+        ps["aNrm"] = s["mNrm"] - a["cNrm"]
+        return ps
+
+    def post_state_transition(self, ps=None, params=None):
+        """
+        post_state to next_state transition
+        """
+        ns = {}  # pytree
+        ns["mNrm"] = ps["aNrm"] * params.Rfree / params.PermGroFac + 1
+        return ns
+
+    def reverse_transition(self, ps=None, a=None, params=None):
+        """state from post_state and actions"""
+        s = {}  # pytree
+        s["mNrm"] = ps["aNrm"] + a["cNrm"]
+
+        return s
+
+    def egm_transition(self, ps=None, continuation=None, params=None):
+        """actions from post_state"""
+
+        a = {}  # pytree
+        a["cNrm"] = self.uP_inv(params.Discount * continuation.derivative(ps))
+
+        return a
+
+    def value_transition(self, action=None, state=None, continuation=None, params=None):
+        """
+        value of action given state and continuation
+        """
+        variables = {}  # pytree
+        ps = self.state_transition(state, action, params)
+        variables.update(ps)
+
+        variables["reward"] = self.u(action["cNrm"])
+        variables["v"] = variables["reward"] + params.Discount * continuation(ps)
```

Comment:
It may be my own ignorance, but the question of whether the application of `params.Discount` should be here or in the continuation (and whether other discounting bits like the PermGroFac term should be here or in the continuation) is confusing to me.

Reply:
For this piece, I wanted to consider `Discount` as a coefficient (float) parameter. Stochastic discounts go into the calculation of the continuation function.